### PR TITLE
[BulkActions] Update BulkActions and add in SelectAllActions

### PR DESCRIPTION
**Pull request by mrcthms**
### WHY are these changes introduced?

Fixes https://github.com/Shopify/web/issues/74192

Spin link in situ: https://admin.web.discounts-update-save-as.marc-thomas.eu.spin.dev/store/shop1/orders?inContextTimeframe=none

#### Before

Screenshot 2022-09-28 at 11 42 54
Screenshot 2022-09-28 at 11 42 28

#### After

Screenshot 2022-09-28 at 11 42 39
Screenshot 2022-09-28 at 11 42 43
Screenshot 2022-09-28 at 11 43 00
Screenshot 2022-09-28 at 11 43 03

This PR updates the BulkActions component, used on the IndexTable and ResourceList components, as well as moving some functionality from BulkActions into the new SelectAllActions component. This is to support the [Unified Index Filtering and Search](https://minerva.shopifycloud.com/auth/minerva?request=2905eaf2-0cf1-493c-b6d3-82e6385d6352&rd=https://vault.shopify.io%2Fprojects%2F24858) project.

This PR introduces a new UX pattern for the BulkActions component. It has been separated from the `Select X` actions – which is now the `SelectAllActions` component, and the `BulkActions` component gets rendered afterward in the DOM.

Components that consume the BulkActions use the new `useIsBulkActionSticky` hook to determine whether the BulkActions component should be `position: fixed` or `position: absolute` with a calculated top position. We have had to do this rather than use a regular `position: sticky` on the BulkActions element, as the keyboard focus order needs to be logically preserved, and moving the BulkActions component under the table/resource list HTML in the DOM would provide a bad experience for users navigating via keyboard.

The IndexTable and ResourceList components have been updated with this new component.

The `CheckableButton` component has been simplified, as a lot of the logic required in the existing implementation is no longer needed. 

As part of this work, small-screen BulkActions and SelectAllActions support has been removed, as we are removing bulk action and select functionality for small screens, as an intentional decision as part of the Unified Index Filtering and Search project.

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [x] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [x] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [x] Updated the component's `README.md` with documentation changes
- [x] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

File: polaris-react/src/components/BulkActions/hooks/use-is-bulk-actions-sticky.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
+import {useEffect, useRef, useState} from 'react';
```
Comment (kyledurand) [inline polaris-react/src/components/BulkActions/hooks/use-is-bulk-actions-sticky.ts:1]: @AndrewMusgrave can you take a look at this hook and where it's used for potential perf improvements? It looks good to me but I think you have more context on resource list and index table.

File: package.json [hunk 2]
```diff
@@ -89,7 +89,7 @@
     {
       "name": "polaris-react-cjs",
       "path": "polaris-react/build/cjs/index.js",
-      "limit": "216 kB"
+      "limit": "218 kB"
```
Comment (AndrewMusgrave) [inline package.json:92]: 📈

File: polaris-react/src/components/ResourceList/ResourceList.tsx [hunk 3]
```diff
@@ -513,15 +528,7 @@ export const ResourceList: ResourceListType = function ResourceList<TItemType>({
...
-    let checkbox: CheckboxHandles | undefined;
-
-    if (isBreakpointsXS()) {
-      checkbox = checkableButtons.get('bulkSm');
-    } else if (newlySelectedItems.length === 0) {
-      checkbox = checkableButtons.get('plain');
-    } else {
-      checkbox = checkableButtons.get('bulkLg');
-    }
+    const checkbox: CheckboxHandles | undefined = checkableButtons.get('plain');
```
Comment (AndrewMusgrave) [inline polaris-react/src/components/ResourceList/ResourceList.tsx]: Do we need a map anymore if `plain` is only checkable button?

File: polaris-react/src/components/ResourceList/ResourceList.scss [hunk 4]
```diff
@@ -155,6 +154,29 @@ $breakpoints-empty-search-results-height-up: '(min-height: #{breakpoint(600px)})
 }
 
 .BulkActionsWrapper {
+  z-index: resource-list(bulk-actions-wrapper-stacking-order);
+  visibility: visible;
```
Comment (AndrewMusgrave) [inline polaris-react/src/components/ResourceList/ResourceList.scss]: Do we ever change the visibility?

File: polaris-react/src/components/CheckableButton/CheckableButton.tsx [hunk 5]
```diff
@@ -14,51 +14,32 @@ export interface CheckableButtonProps {
...
-  } else if (smallScreen) {
-    currentKey = 'bulkSm';
-  }
+  const currentKey: CheckableButtonKey = 'plain';
 
   useEffect(() => {
     if (checkBoxRef.current && registerCheckableButtons) {
       registerCheckableButtons(currentKey, checkBoxRef.current);
     }
   }, [currentKey, registerCheckableButtons]);
```
Comment (AndrewMusgrave) [inline polaris-react/src/components/CheckableButton/CheckableButton.tsx]: We can likely simplify this _a lot_ now and use `useImperativeHandle` exposing a `focus` function

File: polaris-react/src/components/IndexTable/IndexTable.scss [hunk 6]
```diff
@@ -484,14 +488,36 @@ $loading-panel-height: 53px;
 }
 
 .BulkActionsWrapper {
+  visibility: visible;
```
Comment (AndrewMusgrave) [inline polaris-react/src/components/IndexTable/IndexTable.scss:491]: Same visibility comment

File: polaris-react/src/components/BulkActions/BulkActions.scss [hunk 7]
```diff
@@ -7,76 +7,55 @@ $bulk-actions-button-stacking-order: (
...
-  width: 100%;
-  display: none;
+  display: flex;
   align-items: center;
   flex-wrap: wrap;
   opacity: 0;
+  width: 100%;
+  justify-content: center;
+  transition: var(--p-duration-250) var(--p-ease-in);
+  transition-property: transform, opacity;
```
Comment (AndrewMusgrave) [inline polaris-react/src/components/BulkActions/BulkActions.scss:17]: We've seen some issues recently on safari, so we'll want to make sure to thoroughly tophat. `will-change` is something we may want to take advantage of, but we should keep an eye on CPU usage. This is only one element, so I don't suspect it'll to high, however, we've had perf issues in the past with react transition group - so let's keep our eyes open!

File: polaris-react/src/components/BulkActions/BulkActions.scss [hunk 8]
```diff
@@ -7,76 +7,55 @@ $bulk-actions-button-stacking-order: (
...
 }
 
 .ButtonGroupWrapper {
-  width: 100%;
+  width: auto;
+  justify-content: flex-start;
+  padding: var(--p-space-4);
+  background: var(--p-surface);
+  border-radius: var(--p-border-radius-2);
+  box-shadow: var(--p-shadow-popover);
```
Comment (AndrewMusgrave) [inline polaris-react/src/components/BulkActions/BulkActions.scss:57]: ^ (similar comment) - issues recently on safari, however, being only one element I suspect it won't be an issue

File: polaris-react/src/components/BulkActions/hooks/use-is-bulk-actions-sticky.ts [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+      root: null,
+      rootMargin: '0px',
+      threshold: 1,
+    };
+    const observer = new IntersectionObserver(handleIntersect, options);
+
+    const node = bulkActionsIntersectionRef.current;
+
+    if (node) {
+      observer.observe(node);
```
Comment (AndrewMusgrave) [inline polaris-react/src/components/BulkActions/hooks/use-is-bulk-actions-sticky.ts:80]: Is there a lot of overhead in initialization the observer, should it be stored in a ref?

**Review by AndrewMusgrave — COMMENTED** [review]
@kyledurand Scanned the code and left a few comment but nothing large stuck out to me!

Nice work 🙌

File: polaris-react/src/components/CheckableButton/CheckableButton.tsx [hunk 10]
```diff
@@ -14,51 +14,32 @@ export interface CheckableButtonProps {
   accessibilityLabel?: string;
   label?: string;
   selected?: boolean | 'indeterminate';
-  selectMode?: boolean;
-  smallScreen?: boolean;
-  plain?: boolean;
-  measuring?: boolean;
   disabled?: boolean;
   onToggleAll?(): void;
+  ariaLive?: 'off' | 'assertive' | 'polite';
```
Comment (kyledurand) [inline polaris-react/src/components/CheckableButton/CheckableButton.tsx]: I can't imagine we'd ever need `assertive`

File: polaris-react/src/components/SelectAllActions/SelectAllActions.tsx [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+  const checkableButtonProps = {
+    accessibilityLabel,
+    label: hasTextAndAction ? paginatedSelectAllText : label,
+    selected,
+    onToggleAll,
+    disabled,
+    ariaLive,
+  };
+  const markup = (
+    <Transition timeout={0} in={selectMode} key="markup">
```
Comment (kyledurand) [inline polaris-react/src/components/SelectAllActions/SelectAllActions.tsx:77]: I'm guessing this is transitioning `all` properties again when we just need opacity, unless I missed it somewhere. Can we make sure it's specified?

**Review by kyledurand — APPROVED** [review]
Just a couple more minor notes but this looks great to me 👍

File: polaris-react/src/components/IndexTable/IndexTable.scss (hunk 6 above)
Comment (mrcthms) [reply polaris-react/src/components/IndexTable/IndexTable.scss:491]: IT's needed here to negate the `visibility: hidden` of a parent element. Removed it from the `ResourceList.scss` file though, we don't need it there.

File: polaris-react/src/components/CheckableButton/CheckableButton.tsx (hunk 5 above)
Comment (mrcthms) [reply polaris-react/src/components/CheckableButton/CheckableButton.tsx]: Lovely idea, done and removed all the `registerCheckableButtons` code, much simpler/cleaner.

File: polaris-react/src/components/SelectAllActions/SelectAllActions.scss [hunk 12]
```diff
@@ -5,6 +5,7 @@
   gap: var(--p-space-025);
   align-items: center;
   justify-content: flex-start;
+  will-change: opacity, display;
```
Comment (kyledurand) [inline polaris-react/src/components/SelectAllActions/SelectAllActions.scss]: `display` doesn't animate but safari is weird so I get why this _might_ be needed but is it actually needed?

**Pull request merged**